### PR TITLE
Dropdown Autocomplete

### PR DESCRIPTION
**Pull request by alexpaxton**
- [x] CHANGELOG.md updated
  - [x] Rebased/mergable
  - [x] Tests pass

Connect #1462 

### The problem
- Sometimes a dropdown will have a large amount of items in the list, and manually scrolling is frustrating
- Also, there was some confusion between #1462 and the PR #1474 that got merged, so this PR also reverts those changes and then updates them to be consistent with how filtering is now being done app-wide.

### The Solution
- Option for any dropdown to use autocomplete
- Switches to a text input on click, matches as you type
- Can use up/down arrows to change selection
- Esc cancels the action
- Enter confirms selection (if there is one)

**Thanks to @jaredscheib for helping me do the javascripts for this one**

tempvar-autocomplete

### CODE REVIEW

File: ui/src/style/components/dropdown.scss [hunk 1]
```diff
@@ -12,28 +12,47 @@ $dropdown-menu-max-height: 270px;
...
-  &-170 .dropdown-toggle {width: 170px;}
-  &-180 .dropdown-toggle {width: 180px;}
-  &-190 .dropdown-toggle {width: 190px;}
-  &-200 .dropdown-toggle {width: 200px;}
-  &-210 .dropdown-toggle {width: 210px;}
-  &-220 .dropdown-toggle {width: 220px;}
-  &-230 .dropdown-toggle {width: 230px;}
-  &-240 .dropdown-toggle {width: 240px;}
-  &-250 .dropdown-toggle {width: 250px;}
+  &-80 .dropdown-toggle,
```
Comment (lukevmorris) [inline ui/src/style/components/dropdown.scss]: Do we end up using all of these? If so, we can maybe condense using a sass directive:

```sass
@for $i from 8 through 25 {
  &-#{$i * 10} .dropdown-toggle,
  &-#{$i * 10} .dropdown-autocomplete { width: #{$i * 10}px; }
}
```

Comment (alexpaxton) [inline ui/src/style/components/dropdown.scss]: Ahhh excellent call, was thinking about this too

File: ui/src/shared/components/Dropdown.js [hunk 2]
```diff
@@ -54,21 +76,99 @@ class Dropdown extends Component {
...
+      this.props.onChoose(filteredItems[highlightedItemIndex])
+    }
+    if (e.key === 'Escape') {
+      this.setState({isOpen: false})
+    }
+    if (e.key === 'ArrowUp' && highlightedItemIndex > 0) {
+      this.setState({highlightedItemIndex: highlightedItemIndex - 1})
+    }
+    if (
+      e.key === 'ArrowDown' &&
```
Comment (jaredscheib) [inline ui/src/shared/components/Dropdown.js]: nest both further conditions under a single `e.key === 'ArrowDown'

File: ui/src/shared/components/Dropdown.js [hunk 3]
```diff
@@ -54,21 +76,99 @@ class Dropdown extends Component {
     action.handler(item)
   }
 
+  handleFilterKeyPress(e) {
+    const {filteredItems, highlightedItemIndex} = this.state
+
+    if (e.key === 'Enter' && filteredItems.length > 0) {
```
Comment (jaredscheib) [inline ui/src/shared/components/Dropdown.js]: no need for `> 0`

File: ui/src/shared/components/Dropdown.js [hunk 4]
```diff
@@ -164,33 +285,57 @@ class Dropdown extends Component {
...
-        {(isOpen && items.length < DROPDOWN_MENU_ITEM_THRESHOLD)
+        {useAutoComplete && isOpen
+          ? <div
+              className={`dropdown-autocomplete dropdown-toggle ${buttonSize} ${buttonColor}`}
+            >
+              <input
+                ref="dropdownAutoComplete"
+                className="dropdown-autocomplete--input"
+                type="text"
+                autoFocus={true}
```
Comment (jaredscheib) [inline ui/src/shared/components/Dropdown.js]: surprised the linter isn't suggesting to make this simply `autoFocus` -- try it out -- might not need to set it to `true`

File: ui/src/shared/components/Dropdown.js [hunk 5]
```diff
@@ -164,33 +285,57 @@ class Dropdown extends Component {
...
+              <span className="caret" />
+            </div>}
+        {isOpen && menuItems.length < DROPDOWN_MENU_ITEM_THRESHOLD
           ? this.renderShortMenu()
           : null}
-        {(isOpen && items.length >= DROPDOWN_MENU_ITEM_THRESHOLD)
+        {isOpen && menuItems.length >= DROPDOWN_MENU_ITEM_THRESHOLD
           ? this.renderLongMenu()
           : null}
+        {isOpen && menuItems.length === 0
```
Comment (jaredscheib) [inline ui/src/shared/components/Dropdown.js]: could make this `!menuItems.length`

**Comment by jaredscheib** [comment]
Also update the changelog with two fixes, separating out the changes in `6fe8fd1`

**Pull request merged**